Feature/pagination

Sam09040 opened the pull request:
implemented pagination and a bit of restructuring.

Sam09040 committed:
Implemented pagination

- skip arg to users mutation;
- pagination via findMany();
- type UsersPayload to be users query's return;
- totalUsers, page and maxPage values to return on users query;

- prisma variable renamed to dbClient;
- changed a few imports;

Sam09040 committed:
implemented pagination and reestructuring

# added:
- query and mutation division;
- seed client and reset database specific files;
- user related db activities in a separate file;
- password encryption to password.ts;
- pagination implemented to users query;
- tests for users query;

## updated:
- src inner folder name from app to data;
- db client variable name from prisma to dbClient;

walteraandrade commented on src/data/db/user.ts:
```diff
@@ -0,0 +1,31 @@
+    data: {
+      name,
+      email,
+      password: await encryptPassword(password),
```
idealmente esse `encrypt` não deveria estar aqui, nesse arquivo deveríamos apesar usar aquilo que é de responsabilidade do banco de dados, que é guardar as informações

racsandoval commented on src/data/graphql/query/users.query.ts:
```diff
@@ -0,0 +1,48 @@
+
+  const maxPage = Math.round(totalUsers / limit);
+
+  const page = skip ? Math.round(skip / limit) : 0;
```
o `Math.round` arredonda pra cima ou pra baixo, dependendo da casa decimal, nesse caso faz mais sentido utilizar uma função que sempre arredonde pra cima ou para baixo independente da casa decimal

racsandoval commented on src/data/graphql/query/users.query.ts:
```diff
@@ -0,0 +1,48 @@
+
+  const totalUsers = await dbClient.user.count();
+
+  if (!totalUsers) {
+    throw new CustomError('404', 'No users found!', {
+      field: 'User',
+      reason: 'There are no users.',
+    });
+  }
```
por padrão nós retornamos uma array vazia quando não há usuários

Sam09040 replied:
então sem o erro e passando uma array vazia?

racsandoval replied:
isso

racsandoval commented on test/users.test.ts:
```diff
@@ -0,0 +1,141 @@
+  const headers = {
+    'Content-Type': 'application/json',
+    Authorization:
+      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOjEsImlhdCI6MTczMjI4MDc0OSwiZXhwIjoxNzMyODg1NTQ5fQ.huFREPUElN4d9cajXPdngDqSrBDMNDFYvmOVKk9GAMI',
```
colocar o token de autenticação hardcoded aqui não causa problemas no teste? esse token não pode se tornar inválido em algum momento?

Sam09040 replied:
sim, pode causar problemas, ainda mais que expõe o token. vou dar um jeito.

racsandoval commented on test/users.test.ts:
```diff
@@ -0,0 +1,141 @@
+    const response = await axios.post(url, { query, variables }, { headers });
+    const data = response.data.data.users;
+    expect(data).to.have.property('users');
+    expect(data.users.length).to.equal(10);
```
alem de testar a quantidade de usuários, poderia verificar se os campos retornados estão corretos

racsandoval commented on test/users.test.ts:
```diff
@@ -0,0 +1,141 @@
+
+  it('should return correct values for totalUsers, page and maxPage', async () => {
+    await seedUsers(19);
+    const variables = {
+      skip: 10,
+      limit: 0,
+    };
```
poderia testar também mais variações da paginação

racsandoval reviewed (commented):
Foi uma boa ideia fazer a refatoração, ficou bem melhor de ver os arquivos, mas acho que poderia ter feito a paginação e a refatoração em dois PRs separados, ficou difícil avaliar a paginação com tantos arquivos juntos

Sam09040 committed:
users test fixes

## added:
- getToken function to get token for authenticated tests;
- more expects on second test of users test;

## updated:
- headers and token on tests that use it;
- page calculation on users query;
- return empty array if there are no users;
- more variation on tests;

Sam09040 committed:
db fix

## added:
- encryption where it's needed;

## removed:
- encrypt from user and does only db related actions;
